### PR TITLE
Retagging: Recurse into compound values

### PR DESCRIPTION
**Pull request by RalfJung**
This matches https://github.com/rust-lang/rust/pull/55916, waiting for that to land.

### CODE REVIEW

File: src/stacked_borrows.rs [hunk 1]
```diff
@@ -602,17 +602,56 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for MiriEvalContext<'a, 'mir, 'tcx> {
...
+
+        // We need a visitor to visit all references.  However, that requires
+        // a `MemPlace`, so we have a fast path for reference types that
+        // avoids allocating.
+        match place.layout.ty.sty {
+            ty::Ref(_, _, mutbl) => {
+                // fast path
+                let val = self.read_immediate(self.place_to_op(place)?)?;
+                let val = self.reborrow(val, mutbl)?;
+                self.write_immediate(val, place)?;
```
Comment (oli-obk) [inline src/stacked_borrows.rs]: did you forget a `return` here?

File: src/fn_call.rs [hunk 2]
```diff
@@ -315,8 +315,8 @@ impl<'a, 'mir, 'tcx: 'mir + 'a> EvalContextExt<'tcx, 'mir> for super::MiriEvalCo
             }
 
             "memrchr" => {
-                let ptr = self.read_scalar(args[0])?.not_undef()?.erase_tag(); // raw ptr operation
-                let ptr = ptr.with_default_tag();
+                let ptr = self.read_scalar(args[0])?.not_undef()?;
+                let ptr = ptr;
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/fn_call.rs]: Good catch, thanks!

File: src/stacked_borrows.rs (hunk 1 above)
Comment (RalfJung) [reply src/stacked_borrows.rs]: I did, yes. A later commits adds it.

File: src/stacked_borrows.rs [hunk 3]
```diff
@@ -602,17 +602,56 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for MiriEvalContext<'a, 'mir, 'tcx> {
...
+
+            #[inline(always)]
+            fn ecx(&mut self) -> &mut MiriEvalContext<'a, 'mir, 'tcx> {
+                &mut self.ecx
+            }
+
+            // Primitives of reference type, that is the one thing we are interested in.
+            fn visit_primitive(&mut self, place: MPlaceTy<'tcx, Borrow>) -> EvalResult<'tcx>
+            {
+                match place.layout.ty.sty {
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/stacked_borrows.rs]: What exactly?

The problem is that unlike the fast path, this works with an `MPlace`.

File: src/stacked_borrows.rs [hunk 4]
```diff
@@ -661,14 +661,14 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for MiriEvalContext<'a, 'mir, 'tcx> {
...
             {
-                match place.layout.ty.sty {
-                    ty::Ref(_, _, mutbl) => {
-                        let val = self.ecx.read_immediate(place.into())?;
-                        let val = self.ecx.retag_reference(val, mutbl)?;
-                        self.ecx.write_immediate(val, place.into())?;
-                    }
-                    _ => {}, // nothing to do
-                }
+                let mutbl = match place.layout.ty.sty {
```
Comment (oli-obk) [inline src/stacked_borrows.rs]: I think this can be simplified to an if let on `builtin_deref(false)`

Comment (RalfJung) [reply src/stacked_borrows.rs]: Unfortunately that reports *immutable* for `Box`, making it useless.

File: src/stacked_borrows.rs [hunk 5]
```diff
@@ -602,17 +618,57 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for MiriEvalContext<'a, 'mir, 'tcx> {
...
-        // TODO: Recurse into structs and enums, sharing code with validation.
         // TODO: Honor `fn_entry`.
-        let mutbl = match place.layout.ty.sty {
-            ty::Ref(_, _, mutbl) => mutbl, // go ahead
-            _ => return Ok(()), // do nothing, for now
+
+        // We need a visitor to visit all references.  However, that requires
+        // a `MemPlace`, so we have a fast path for reference types that
+        // avoids allocating.
+        match place.layout.ty.sty {
```
Comment (oli-obk) [inline src/stacked_borrows.rs]: Does this fast path also work for boxes?

File: src/stacked_borrows.rs (hunk 4 above)
Comment (RalfJung) [reply src/stacked_borrows.rs]: I added a comment explaining this.

File: src/stacked_borrows.rs (hunk 5 above)
Comment (RalfJung) [reply src/stacked_borrows.rs]: Ah, I guess it should. I was just lazy.^^

Comment (RalfJung) [reply src/stacked_borrows.rs]: Done.

File: src/stacked_borrows.rs (hunk 4 above)
Comment (oli-obk) [reply src/stacked_borrows.rs]: Wat. I need to review all uses of that method now -.-

Comment (RalfJung) [reply src/stacked_borrows.rs]: Please do.^^ And/or fix that method? ;)

**Comment by RalfJung** [comment]
Bumped Rust version and merged https://github.com/solson/miri/pull/523, this should work now.

**Pull request merged**